pkg/platform/api/buildplanner/model/buildexpression.go

```diff
@@ -97,122 +167,254 @@ type BuildExpression struct {
-	for i, req := range bx.requirements {
-		if req.Name == requirement.Name && req.Namespace == requirement.Namespace {
-			bx.requirements = append(bx.requirements[:i], bx.requirements[i+1:]...)
-			(*bx.solveNode)[RequirementsKey] = bx.requirements
-			return nil
+func newAp(ctx context, m map[string]interface{}) (*Ap, error) {
+	ctx.push(ctxAp)
+	defer ctx.pop()
+
+	// Look in the given object for the function's name and argument object or list.
+	var name string
+	var argsInterface interface{}
+	for key, value := range m {
+		if isAp(ctx, value.(map[string]interface{})) {
+			name = key
+			argsInterface = value
+			break
 		}
 	}
 
-	return errs.New("Could not find requirement")
+	args := []*Value{}
+
+	switch v := argsInterface.(type) {
+	case map[string]interface{}:
+		for key, valueInterface := range v {
+			value, err := newValue(ctx, valueInterface)
+			if err != nil {
+				return nil, errs.Wrap(err, "Could not parse '%s' function's argument '%s': %v", name, key, valueInterface)
+			}
+			args = append(args, &Value{Assignment: &Var{Name: key, Value: value}})
+		}
+		sort.SliceStable(args, func(i, j int) bool { return args[i].Assignment.Name < args[j].Assignment.Name })
+
+	case []interface{}:
+		for _, item := range v {
+			value, err := newValue(ctx, item)
+			if err != nil {
+				return nil, errs.Wrap(err, "Could not parse '%s' function's argument list item: %v", name, item)
+			}
+			args = append(args, value)
+		}
+
+	default:
+		return nil, errs.New("Function '%s' expected to be object or list", name)
+	}
+
+	return &Ap{Name: name, Arguments: args}, nil
 }
 
-// updateRequirement updates an existing requirement in the BuildExpression.
-func (bx BuildExpression) updateRequirement(requirement Requirement) error {
-	for i, req := range bx.requirements {
-		if req.Name == requirement.Name && req.Namespace == requirement.Namespace {
-			bx.requirements[i] = requirement
-			(*bx.solveNode)[RequirementsKey] = bx.requirements
-			return nil
+func newAssignments(ctx context, m map[string]interface{}) (*[]*Var, error) {
+	ctx.push(ctxAssignments)
+	defer ctx.pop()
+
+	assignments := []*Var{}
+	for key, valueInterface := range m {
+		value, err := newValue(ctx, valueInterface)
+		if err != nil {
+			return nil, errs.Wrap(err, "Could not parse '%s' key's value: %v", key, valueInterface)
 		}
+		assignments = append(assignments, &Var{Name: key, Value: value})
 	}
+	sort.SliceStable(assignments, func(i, j int) bool { return assignments[i].Name < assignments[j].Name })
+	return &assignments, nil
+}
+
+func newIn(ctx context, inValue interface{}) (*In, error) {
+	ctx.push(ctxIn)
+	defer ctx.pop()
+
+	in := &In{}
+
+	switch v := inValue.(type) {
+	case map[string]interface{}:
+		f, err := newAp(ctx, v)
+		if err != nil {
+			return nil, errs.Wrap(err, "'in' object is not a function call")
+		}
+		in.FuncCall = f
+
+	case string:
+		in.Name = p.StrP(strings.TrimPrefix(v, "$"))
```

Comment:
Out of curiosity, what is '$' for? I couldn't figure out what '$' was doing, so I was stripping it/adding it as necessary when processing build scripts without actually knowing if what I was doing was correct.

Reply:
It's to reference a variable. When `runtime` is assigned it's not there but for any reference, it should be.